### PR TITLE
feat(server): introduce browser server mode

### PR DESCRIPTION
**Pull request by pavelfeldman**

### CODE REVIEW

File: src/playwrightTest.ts [hunk 1]
```diff
@@ -222,6 +229,52 @@ export class PlaywrightTest {
...
+        return;
+      this._browserServerWS = match[1];
+    });
+
+    stdio[2].on('data', data => {
+      console.log('Error in browser server: ' + data.toString());
+    });
+
+    token.onCancellationRequested(() => {
+      serverProcess.stdin.write('<EOL>');
```
Comment (dgozman) [inline src/playwrightTest.ts:263]: I think we used to `stdin.close()` in other places instead? That would just imitate disconnect.

File: src/playwrightTest.ts [hunk 2]
```diff
@@ -241,6 +294,17 @@ export class PlaywrightTest {
     return node;
   }
 
+  async findNpx(): Promise<string> {
+    if (this._pathToNpxJS)
+      return this._pathToNpxJS;
+
+    const npx = await which('npx');
+    if (!npx)
+      throw new Error('Unable to launch `node`, make sure it is in your PATH');
```
Comment (dgozman) [inline src/playwrightTest.ts:303]:
```suggestion
      throw new Error('Unable to launch `npx`, make sure it is in your PATH');
```

File: src/playwrightTest.ts [hunk 3]
```diff
@@ -222,6 +229,52 @@ export class PlaywrightTest {
     await reporterServer.wireTestListener(listener, token);
   }
 
+  async runBrowserServer(workspaceFolder: string, token: vscodeTypes.CancellationToken) {
+    if (this._browserServerWS)
+      return;
+
+    const npx = await this.findNpx();
+
+    const allArgs = [
```
Comment (dgozman) [inline src/playwrightTest.ts:238]: If you do `node playwright-core/cli.js ...`, that will save a process and a second.

File: src/playwrightTest.ts [hunk 4]
```diff
@@ -149,17 +151,22 @@ export class PlaywrightTest {
       '--repeat-each', '1',
       '--retries', '0',
     ];
-    if (this._isUnderTest)
+    if (this._isUnderTest || !!this._browserServerWS)
```
Comment (dgozman) [inline src/playwrightTest.ts:154]: Should we check that playwright versions from `config` and `runBrowserServer` match and error otherwise?

**Review by dgozman — APPROVED** [review]

File: src/playwrightTest.ts (hunk 1 above)
Comment (pavelfeldman) [reply src/playwrightTest.ts:263]: Unfortunately, it does not. I think we were doing it in our Java port only.

File: src/playwrightTest.ts (hunk 4 above)
Comment (pavelfeldman) [reply src/playwrightTest.ts:154]: That would mean that we have a multi-workspace configuration in VS Code or a monorepo where different projects have different versions of Playwright. I'd say ignore this edge case for now? I should add a minimum version check for run-server though, I'll do that.

File: src/playwrightTest.ts (hunk 3 above)
Comment (pavelfeldman) [reply src/playwrightTest.ts:237]: I originally was trying to implement it in a model-less manner, but I guess it is ok.

**Pull request merged**